### PR TITLE
bug 1792386: set the operand version and make the CVO wait

### PR DESCRIPTION
**Pull request by deads2k**
builds on #301 

The last commit is unique and sets the operand to the same version as the operator when it has achieved its consistent level.

/assign @mfojtik

### CODE REVIEW

**Comment by deads2k** [comment]
```
  versions:
  - name: operator
    version: 0.0.1-2020-01-17-191329
  - name: openshift-apiserver
    version: 0.0.1-2020-01-17-191329
```

set in two go-s
```
  message: clusteroperator/openshift-apiserver version "operator" changed from "0.0.1-2020-01-17-191141"
    to "0.0.1-2020-01-17-191329"
  metadata:
    creationTimestamp: "2020-01-17T20:00:46Z"

message: clusteroperator/openshift-apiserver version "openshift-apiserver" changed
    from "" to "0.0.1-2020-01-17-191329"
  metadata:
    creationTimestamp: "2020-01-17T20:04:27Z"
```


four minutes later after the daemonset stabilized post-upgrade.

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 1]
```diff
@@ -75,116 +77,173 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+		}))...,
+	)
+	errors = append(errors,
+		appendErrors(v1helpers.UpdateStatus(c.operatorClient, func(status *operatorv1.OperatorStatus) error {
+			resourcemerge.SetDaemonSetGeneration(&status.Generations, actualDaemonSet)
+			return nil
+		}))...,
+	)
+
+	daemonSetHasAllPodsAvailable := actualDaemonSet.Status.NumberAvailable == actualDaemonSet.Status.DesiredNumberScheduled
```
Comment (p0lyn0mial) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: what's the difference between `NumberAvailable ` and `UpdatedNumberScheduled`?

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 2]
```diff
@@ -75,116 +77,173 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+	if !daemonSetHasAllPodsAvailable {
+		numNonAvailablePods := actualDaemonSet.Status.DesiredNumberScheduled - actualDaemonSet.Status.NumberAvailable
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "APIServerDaemonSetDegraded",
```
Comment (p0lyn0mial) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: yeah, if `daemonSetAtHighestGeneration` then it is degraded

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go (hunk 1 above)
Comment (mfojtik) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: This is consistent with `kubectl` rollout status:

```go
112:		if daemon.Status.NumberAvailable < daemon.Status.DesiredNumberScheduled {
113:			return fmt.Sprintf("Waiting for daemon set %q rollout to finish: %d of %d updated pods are available...\n", daemon.Name, daemon.Status.NumberAvailable, daemon.Status.DesiredNumberScheduled), false, nil
```

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 3]
```diff
@@ -75,116 +75,169 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+				Type:    "OperatorConfigProgressing",
+				Status:  operatorv1.ConditionTrue,
+				Reason:  "NewGeneration",
+				Message: fmt.Sprintf("openshiftapiserveroperatorconfigs/instance: observed generation is %d, desired generation is %d.", operatorConfig.Status.ObservedGeneration, operatorConfig.ObjectMeta.Generation),
+			})))...,
+		)
+	} else {
+		errors = append(errors,
+			appendErrors(v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
```
Comment (p0lyn0mial) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: wouldn’t be better to collect all the updates and update only once at the end ?

**Pull request merged**